janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java

```diff
@@ -432,26 +432,28 @@ private static Sort getSortOrder(IndexQuery query) {
     }
 
     @Override
-    public List<String> query(IndexQuery query, KeyInformation.IndexRetriever informations, BaseTransaction tx) throws BackendException {
+    public Stream<String> query(IndexQuery query, KeyInformation.IndexRetriever informations, BaseTransaction tx) throws BackendException {
         //Construct query
         SearchParams searchParams = convertQuery(query.getCondition(),informations.get(query.getStore()));
 
         try {
             IndexSearcher searcher = ((Transaction) tx).getSearcher(query.getStore());
-            if (searcher == null) return ImmutableList.of(); //Index does not yet exist
+            if (searcher == null) return Collections.unmodifiableList(new ArrayList<String>()).stream(); //Index does not yet exist
 
             Query q = searchParams.getQuery();
             if (null == q)
                 q = new MatchAllDocsQuery();
 
             long time = System.currentTimeMillis();
-            final TopDocs docs = searcher.search(q, query.hasLimit() ? query.getLimit() : Integer.MAX_VALUE - 1, getSortOrder(query));
+            TopDocs docs = searcher.search(q, query.hasLimit() ? query.getLimit() : Integer.MAX_VALUE - 1, getSortOrder(query));
             log.debug("Executed query [{}] in {} ms", q, System.currentTimeMillis() - time);
-            List<String> result = new ArrayList<String>(docs.scoreDocs.length);
-            for (int i = 0; i < docs.scoreDocs.length; i++) {
-                result.add(searcher.doc(docs.scoreDocs[i].doc).getField(DOCID).stringValue());
-            }
-            return result;
+            return Arrays.stream(docs.scoreDocs).map(match -> {
+                try {
+                    return searcher.doc(match.doc).getField(DOCID).stringValue();
+                } catch (IOException e) {
+                    throw new RuntimeException("Could not execute Lucene query", e);
```

Comment:
In this case it might be worth sticking with a non-stream implementation so the IOException can be thrown and caught below for consistent exception handling (e.g. TemporaryBackendException) in the method.